File: example/kotlinlib/web/5-webapp-kotlinjs-shared/shared/src/Shared.kt
package shared

import kotlinx.html.FlowContent
import kotlinx.html.InputType
import kotlinx.html.a
import kotlinx.html.button
import kotlinx.html.div
import kotlinx.html.footer
import kotlinx.html.h1
import kotlinx.html.header
import kotlinx.html.id
import kotlinx.html.input
import kotlinx.html.label
import kotlinx.html.li
import kotlinx.html.section
import kotlinx.html.span
import kotlinx.html.strong
import kotlinx.html.ul
import kotlinx.serialization.Serializable

@Serializable
data class Todo(
    val checked: Boolean,
    val text: String,
)

fun FlowContent.renderBody(
    todos: List<Todo>,
    state: String,
) {
    val filteredTodos =
        when (state) {
            "all" -> todos.withIndex()
            "active" -> todos.withIndex().filter { !it.value.checked }
            "completed" -> todos.withIndex().filter { it.value.checked }
            else -> throw IllegalStateException("Unknown state=$state")
        }
    div {
        header(classes = "header") {
            h1 { +"todos" }
            input(classes = "new-todo") {
                placeholder = "What needs to be done?"
            }
        }
        section(classes = "main") {
            input(
                classes = "toggle-all",
                type = InputType.checkBox,
            ) {
                id = "toggle-all"
                checked = todos.any { it.checked }
            }
            label {
                htmlFor = "toggle-all"
                +"Mark all as complete"
            }
            ul(classes = "todo-list") {
                filteredTodos.forEach { (index, todo) ->
                    li(classes = if (todo.checked) "completed" else "") {
                        div(classes = "view") {
                            input(classes = "toggle", type = InputType.checkBox) {
                                checked = todo.checked
                                attributes["data-todo-index"] = index.toString()
                            }
                            label { +todo.text }
                            button(classes = "destroy") {
                                attributes["data-todo-index"] = index.toString()
                            }
                        }
                        input(classes = "edit") {
                            value = todo.text
                        }
                    }
                }
            }
        }
        footer(classes = "footer") {
            span(classes = "todo-count") {
                strong {
                    +todos.filter { !it.checked }.size.toString()
                }
                +" items left"
            }
            ul(classes = "filters") {
                li(classes = "todo-all") {
                    a(classes = if (state == "all") "selected" else "") { +"All" }
                }
                li(classes = "todo-active") {
                    a(classes = if (state == "active") "selected" else "") { +"Active" }
                }
                li(classes = "todo-completed") {
                    a(classes = if (state == "completed") "selected" else "") { +"Completed" }
                }
            }
            button(classes = "clear-completed") { +"Clear completed" }
        }
    }
}

File: example/scalalib/web/6-webapp-scalajs-shared/shared/src/Shared.scala
package shared
import scalatags.Text.all.*
import scalatags.Text.tags2

case class Todo(checked: Boolean, text: String)

object Todo {
  implicit def todoRW: upickle.default.ReadWriter[Todo] = upickle.default.macroRW[Todo]
}

object Shared {
  def renderBody(todos: Seq[Todo], state: String) = {
    val filteredTodos = state match {
      case "all" => todos.zipWithIndex
      case "active" => todos.zipWithIndex.filter(!_._1.checked)
      case "completed" => todos.zipWithIndex.filter(_._1.checked)
    }
    div(
      header(cls := "header")(
        h1("todos"),
        input(cls := "new-todo", placeholder := "What needs to be done?", autofocus := "")
      ),
      tags2.section(cls := "main")(
        input(
          id := "toggle-all",
          cls := "toggle-all",
          `type` := "checkbox",
          if (todos.filter(_.checked).size != 0) checked else ()
        ),
        label(`for` := "toggle-all")("Mark all as complete"),
        ul(cls := "todo-list")(
          for ((todo, index) <- filteredTodos) yield li(
            if (todo.checked) cls := "completed" else (),
            div(cls := "view")(
              input(
                cls := "toggle",
                `type` := "checkbox",
                if (todo.checked) checked else (),
                data("todo-index") := index
              ),
              label(todo.text),
              button(cls := "destroy", data("todo-index") := index)
            ),
            input(cls := "edit", value := todo.text)
          )
        )
      ),
      footer(cls := "footer")(
        span(cls := "todo-count")(
          strong(todos.filter(!_.checked).size),
          " items left"
        ),
        ul(cls := "filters")(
          li(cls := "todo-all")(
            a(if (state == "all") cls := "selected" else ())("All")
          ),
          li(cls := "todo-active")(
            a(if (state == "active") cls := "selected" else ())("Active")
          ),
          li(cls := "todo-completed")(
            a(if (state == "completed") cls := "selected" else ())("Completed")
          )
        ),
        button(cls := "clear-completed")("Clear completed")
      )
    )
  }

}
